src/Balea/Authorization/BaleaPolicyEvaluator.cs

```diff
@@ -77,14 +83,68 @@ public override async Task<AuthenticateResult> AuthenticateAsync(AuthorizationPo
             return AuthenticateResult.NoResult();
         }
 
+        public async Task<PolicyAuthorizationResult> AuthorizeAsync(
+            AuthorizationPolicy policy,
+            AuthenticateResult authenticationResult,
+            HttpContext context,
+            object resource)
+        {
+            if (policy == null)
+            {
+                throw new ArgumentNullException(nameof(policy));
+            }
+
+            var result = await _authorization.AuthorizeAsync(context.User, resource, policy);
+
+            if (result.Succeeded)
+            {
+                _logger.PolicySucceed();
+                return PolicyAuthorizationResult.Success();
+            }
+
+            // If authentication was successful, return forbidden, otherwise challenge
+            if (authenticationResult.Succeeded)
+            {
+                _logger.PolicyFailToForbid();
+                return PolicyAuthorizationResult.Forbid();
+            }
+
+            _logger.PolicyFailToChallenge();
+            return PolicyAuthorizationResult.Challenge();
+        }
+
         private async Task AddBaleaIdentity(ClaimsPrincipal user, HttpContext context)
         {
             var authorization = await _store
                 .FindAuthorizationAsync(user);
 
-            if (!context.Response.HasStarted && _options.UnauthorizedFallback != null && !authorization.Roles.Any())
+            if (authorization.Roles.Any())
```

Comment:
@lurumad here I've changed slightly the original behavior.
Does it make sense?

Reply:
Does make sense ;)